iiod/responder.c

```diff
@@ -367,23 +367,53 @@ static void handle_create_buffer(struct parser_pdata *pdata,
 	struct iio_channels_mask *mask;
 	const struct iio_device *dev;
 	struct iio_channel *chn;
-	struct buffer_entry *entry;
+	struct buffer_entry *entry, *existing_entry, *placeholder_entry;
 	struct iio_buffer *buf;
 	struct iiod_buf_params *params;
 	struct iiod_buf data;
  	struct iio_buffer_params buffer_params = {0};
 	unsigned int i, nb_channels;
 	size_t nb_words;
+	uint16_t buffer_idx = (uint16_t) cmd->code;
 	int ret = -EINVAL;
 
 	dev = iio_context_get_device(ctx, cmd->dev);
 	if (!dev)
 		goto err_send_response;
 
+	/* Create placeholder entry to reserve the slot */
+	placeholder_entry = zalloc(sizeof(*placeholder_entry));
+	if (!placeholder_entry) {
+		ret = -ENOMEM;
+		goto err_send_response;
+	}
+
+	placeholder_entry->dev = dev;
+	placeholder_entry->idx = buffer_idx;
+	placeholder_entry->buf = NULL; /* Mark as placeholder */
+	placeholder_entry->pdata = pdata;
+
+	/* Check atomically if buffer with same device and index already exists,
+	 * and if not, insert placeholder to reserve the slot */
+	iio_mutex_lock(buflist_lock);
+
+	SLIST_FOREACH(existing_entry, &bufferlist, entry) {
+		if (existing_entry->dev == dev && existing_entry->idx == buffer_idx) {
+			iio_mutex_unlock(buflist_lock);
+			free(placeholder_entry);
+			ret = -EEXIST;
+			goto err_send_response;
+		}
+	}
```

Comment:
I'm not sure all this dance is worth it in comparison with locking the whole thing. I mean, I think that until after `iio_create_channels_mask()` should be thread safe. After that point we start already configuring the device so I would say  to lock the `buflist_lock` after that call. Then, you could use `get_iio_block_unlocked()` to check for the buffer existence and act accordingly. I think it should work and IMO, it's much simpler. I also dunno for any contention issue (and this should not be a fast path anyways so I would go for simplicity)